File: src/Platforms/Android/ConfigImportReceiver.cs
using Android.App;
using Android.Content;
using Android.OS;
using CommunityToolkit.Mvvm.Messaging;
using Expandroid.Models;
using System.Text.Json;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

[BroadcastReceiver(Enabled = true, Exported = true, Name = "com.dingleinc.texttoolspro.ConfigImportReceiver")]
[IntentFilter(new[] { "com.dingleinc.texttoolspro.IMPORT_CONFIG" })]
public class ConfigImportReceiver : BroadcastReceiver
{
    //// Tasker and MacroDroid package names
    //private static readonly string[] AllowedPackages = new[]
    //{
    //    "net.dinglisch.android.taskerm",    // Tasker
    //    "com.arlosoft.macrodroid",           // MacroDroid
    //    "com.dingleinc.texttoolspro"
    //};
    private void SendMessage(string cmd, Match value)
    {
        WeakReferenceMessenger.Default.Send(new AcServiceMessage((cmd, value)));
    }
    public override void OnReceive(Context? context, Intent? intent)
    {
        if (intent is null) return;
        // Only check on Android 4.4+ (API 19+)
        if (Build.VERSION.SdkInt >= BuildVersionCodes.JellyBeanMr1)
        {
            //int callingUid = Binder.CallingUid;
            //var pm = context.PackageManager;
            //var packages = pm.GetPackagesForUid(callingUid);
            //if (packages == null || !packages.Any(pkg => AllowedPackages.Contains(pkg)))
            //    return;
            var configStr = intent.GetStringExtra("config_string");
            if (!string.IsNullOrEmpty(configStr))
            {
                try
                {
                    var deserializer = new DeserializerBuilder()
                                .WithNamingConvention(UnderscoredNamingConvention.Instance).IgnoreUnmatchedProperties()
                                .Build();
                    var localDict = deserializer.Deserialize<DictWrapper>(configStr);
                    foreach (var item in localDict.Matches)
                    {
                        if (item.Vars is not null)
                        {
                            bool notSupported = item.Replace is null;
                            foreach (var x in item.Vars)
                            {
                                if (x.Type is not null)
                                {
                                    if (!AppSettings.SupportedList.Contains(x.Type))
                                    {
                                        notSupported = true;
                                        Android.Util.Log.Warn("ConfigImport", $"Skipped match '{item.Trigger}': unsupported var type '{x.Type}'");
                                        break;
                                    }
                                    else if (x.Type == "date")
                                    {
                                        try
                                        {
                                            x.Params.Format = Utils.GetTheRealFormat(x.Params.Format);
                                        }
                                        catch (Exception)
                                        {
                                            throw new Exception("Date extension parameter formats error");
                                        }
                                    }
                                }
                            }
                            if (notSupported)
                            {
                                if (item.Replace is null)
                                    Android.Util.Log.Warn("ConfigImport", $"Skipped match '{item.Trigger}': missing replace field");
                                continue;
                            }
                        }
                    }
                    if (localDict.Global_vars is not null)
                    {
                        var str = JsonSerializer.Serialize(localDict.Global_vars);
                        File.WriteAllText(AppSettings.GlobalVarsPath, str);
                        WeakReferenceMessenger.Default.Send(new AcGlobalsMessage(localDict.Global_vars));
                    }
                    Dictionary<string, Match> dict = new();
                    foreach (var match in localDict.Matches)
                    {
                        if (match.Triggers is not null && match.Triggers.Count > 0)
                        {
                            foreach (var t in match.Triggers)
                            {
                                var clone = new Match(match) { Trigger = t };
                                dict[t] = clone;
                            }
                        }
                        else if (!string.IsNullOrEmpty(match.Trigger))
                        {
                            dict[match.Trigger] = match;
                        }
                        else if (!string.IsNullOrEmpty(match.Regex))
                        {
                            dict[$"__regex_{match.Regex}"] = match;
                        }
                    }
                    var jsonStr = JsonSerializer.Serialize(dict);
                    File.WriteAllText(AppSettings.DictPath, jsonStr);
                    SendMessage("Reset", new Match());
                    Intent resultIntent = new Intent("com.dingleinc.texttoolspro.CONFIG_RESULT");
                    resultIntent.PutExtra("status", 0); // or 1 for failure
                    context.SendBroadcast(resultIntent);

                }
                catch (Exception e)
                {
                    Intent resultIntent = new Intent("com.dingleinc.texttoolspro.CONFIG_RESULT");
                    resultIntent.PutExtra("status", e.Message); // or 1 for failure
                    context.SendBroadcast(resultIntent);
                }
            }

        }
    }
}

File: app/src/main/java/com/dingleinc/texttoolspro/service/ConfigImportReceiver.kt
package com.dingleinc.texttoolspro.service

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import android.util.Log
import com.dingleinc.texttoolspro.data.AppSettings
import com.dingleinc.texttoolspro.data.Match
import com.dingleinc.texttoolspro.data.SerializationHelper
import com.dingleinc.texttoolspro.data.ServiceCommandBus
import com.dingleinc.texttoolspro.data.Var
import com.dingleinc.texttoolspro.util.Utils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import java.io.File

class ConfigImportReceiver : BroadcastReceiver() {

    companion object {
        private const val TAG = "ConfigImport"
    }

    override fun onReceive(context: Context, intent: Intent) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR1) return

        val configStr = intent.getStringExtra("config_string") ?: return
        if (configStr.isEmpty()) return

        val pendingResult = goAsync()
        CoroutineScope(SupervisorJob() + Dispatchers.IO).launch {
            try {
                val localDict = SerializationHelper.parseDictWrapperFromYaml(configStr)

                localDict.matches?.forEach { item ->
                    item.vars?.forEach { x ->
                        if (x.type != null) {
                            if (!AppSettings.supportedList.contains(x.type)) {
                                return@forEach
                            } else if (x.type == "date") {
                                try {
                                    val fmt = x.params.string("format")
                                    if (!fmt.isNullOrEmpty()) {
                                        x.params["format"] = Utils.getTheRealFormat(fmt)
                                    }
                                } catch (e: Exception) {
                                    throw Exception("Date extension parameter formats error")
                                }
                            }
                        }
                    }
                }

                localDict.globalVars?.let { gvars ->
                    val str = SerializationHelper.toJson(gvars)
                    File(AppSettings.globalVarsPath).writeText(str)
                    ServiceCommandBus.trySend(
                        ServiceCommandBus.Command.UpdateGlobals(gvars)
                    )
                }

                val dict = mutableMapOf<String, Match>()
                var skippedCount = 0
                localDict.matches?.forEach { match ->
                    // Check if any vars have unsupported types
                    var hasUnsupported = false
                    match.vars?.forEach { v ->
                        if (v.type != null && !AppSettings.supportedList.contains(v.type)) {
                            hasUnsupported = true
                        }
                    }
                    if (hasUnsupported) {
                        skippedCount++
                        Log.w(TAG, "Skipped match with unsupported var type: trigger=${match.trigger} regex=${match.regex}")
                        return@forEach
                    }

                    if (!match.triggers.isNullOrEmpty()) {
                        match.triggers!!.forEach { t ->
                            val cloned = Match(match)
                            cloned.trigger = t
                            cloned.triggers = null
                            dict[t] = cloned
                        }
                    } else if (!match.trigger.isNullOrEmpty()) {
                        dict[match.trigger!!] = match
                    } else if (!match.regex.isNullOrEmpty()) {
                        // Regex-only match: use regex string as dict key for storage
                        dict["__regex__${match.regex}"] = match
                    }
                }
                val jsonStr = SerializationHelper.toJson(dict)
                File(AppSettings.dictPath).writeText(jsonStr)
                ServiceCommandBus.trySend(ServiceCommandBus.Command.Reset)

                val resultIntent = Intent("com.dingleinc.texttoolspro.CONFIG_RESULT")
                resultIntent.putExtra("status", 0)
                resultIntent.putExtra("imported", dict.size)
                resultIntent.putExtra("skipped", skippedCount)
                context.sendBroadcast(resultIntent)
            } catch (e: Exception) {
                Log.e(TAG, "Import error: ${e}")
                val resultIntent = Intent("com.dingleinc.texttoolspro.CONFIG_RESULT")
                resultIntent.putExtra("status", e.message ?: "error")
                context.sendBroadcast(resultIntent)
            } finally {
                pendingResult.finish()
            }
        }
    }
}
